src/main/java/net/neoforged/neoforge/common/crafting/outgredient/TagDefaultsManager.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting.outgredient;
+
+import com.google.gson.JsonElement;
+import com.mojang.logging.LogUtils;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.JsonOps;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import net.minecraft.core.Holder;
+import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.FileToIdConverter;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
+import net.minecraft.tags.TagKey;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.neoforged.neoforge.common.NeoForgeEventHandler;
+import net.neoforged.neoforge.common.config.NeoForgeServerConfig;
+import org.slf4j.Logger;
+
+/**
+ * This is the central manager class for {@link TagDefaults}. Access an instance via {@link NeoForgeEventHandler#getTagDefaultsManager()}.
+ */
+public class TagDefaultsManager extends SimpleJsonResourceReloadListener<JsonElement> {
+    public static final String FOLDER = "tag_defaults";
+    private static final Logger LOGGER = LogUtils.getLogger();
+    private final Map<ResourceKey<? extends Registry<?>>, TagDefaults<?>> entries = new HashMap<>();
+
+    public TagDefaultsManager() {
+        super(ExtraCodecs.JSON, FileToIdConverter.json(FOLDER));
+    }
+
+    @Override
+    protected void apply(Map<ResourceLocation, JsonElement> resourceList, ResourceManager resourceManager, ProfilerFiller profilerFiller) {
+        entries.clear();
+        resourceList.forEach(this::put);
+    }
+
+    /**
+     * Resolves a {@link TagDefaults} into an optional outgredient.
+     *
+     * @param registryKey The key of the associated {@link Registry}.
+     * @param tagKey      The {@link TagKey} to resolve.
+     * @return An {@link Optional} of a {@link TagDefaults}'s resolved value, or an empty {@link Optional} if the {@link TagDefaults} could not be resolved.
+     * @param <T> The type of the {@link Registry} and the outgredient.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> resolve(ResourceKey<? extends Registry<T>> registryKey, TagKey<T> tagKey) {
```

Comment:
You know `TagKey<T>` contains its registry through `tagKey.registry()`? No need to pass it separately.